source/grid/grid_in.cc

```diff
@@ -64,35 +64,41 @@ namespace
    * boundary indicators on vertices after the triangulation has already been
    * created.
    *
-   * TODO: Fix this properly via SubcellData
+   * This function ignores all non-boundary vertices. Boundary ids can only be
+   * assigned to internal vertices - however, the grid fixup routines (e.g.,
+   * combining duplicated vertices) may switch vertices from boundary to
+   * non-boundary. Get around this in two ways:
+   *
+   * 1. Ignore any attempt to set a boundary id on a non-boundary vertex.
+   * 2. Work around possible vertex renumberings by using Point<spacedim>
+   *   instead of the corresponding vertex number (which may be out of date)
+   *
+   * TODO: fix this properly via SubCellData
    */
   template <int spacedim>
   void
   assign_1d_boundary_ids(
-    const std::map<unsigned int, types::boundary_id> &boundary_ids,
-    Triangulation<1, spacedim> &                      triangulation)
+    const std::vector<std::pair<Point<spacedim>, types::boundary_id>>
+      &                         boundary_ids,
+    Triangulation<1, spacedim> &triangulation)
   {
-    if (boundary_ids.size() > 0)
-      for (const auto &cell : triangulation.active_cell_iterators())
-        for (unsigned int f : GeometryInfo<1>::face_indices())
-          if (boundary_ids.find(cell->vertex_index(f)) != boundary_ids.end())
-            {
-              AssertThrow(
-                cell->at_boundary(f),
-                ExcMessage(
-                  "You are trying to prescribe boundary ids on the face "
-                  "of a 1d cell (i.e., on a vertex), but this face is not actually at "
-                  "the boundary of the mesh. This is not allowed."));
-              cell->face(f)->set_boundary_id(
-                boundary_ids.find(cell->vertex_index(f))->second);
-            }
+    for (auto &cell : triangulation.active_cell_iterators())
+      for (unsigned int face_no : ReferenceCells::Line.face_indices())
+        if (cell->face(face_no)->at_boundary())
+          for (const auto &pair : boundary_ids)
```

Comment:
For a moment I thought "oh, that's expensive". But you're only doing the O(N) loop twice for 1d domains, so it doesn't matter.

Reply:
It might be more than once if codim = 1 or codim = 2 but I don't think it's ever going to be a noticeably large computational cost. We can always go back and sort the vertices if we need to.